src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -545,24 +545,38 @@ void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
         scheduling_key_entry.pending_lease_requests.erase(task_id);
 
         if (status.ok()) {
-          if (reply.runtime_env_setup_failed()) {
-            // If the runtime_env failed to be set up, we fail all of the pending
-            // tasks in the queue. This makes an implicit assumption that runtime_env
-            // failures are not transient -- we may consider adding some retries
-            // in the future.
-            auto &task_queue = scheduling_key_entry.task_queue;
-            while (!task_queue.empty()) {
-              auto &task_spec = task_queue.front();
-              RAY_UNUSED(task_finisher_->MarkPendingTaskFailed(
-                  task_spec, rpc::ErrorType::RUNTIME_ENV_SETUP_FAILED, nullptr));
-              task_queue.pop_front();
-            }
-            if (scheduling_key_entry.CanDelete()) {
-              scheduling_key_entries_.erase(scheduling_key);
+          if (reply.canceled()) {
+            RAY_LOG(DEBUG) << "Lease canceled for task: " << task_id << ", cancel type: "
+                           << rpc::RequestWorkerLeaseReply::CancelType_Name(
+                                  reply.cancel_type());
+            if (reply.cancel_type() ==
+                    rpc::RequestWorkerLeaseReply::RUNTIME_ENV_SETUP_FAILED ||
+                reply.cancel_type() ==
+                    rpc::RequestWorkerLeaseReply::PLACEMENT_GROUP_REMOVED) {
+              // If the runtime_env failed to be set up, we fail all of the pending
```

Comment:
This comment seems to be wrong. So, the behavior is we cancel all pending tasks when pg is removed right?

Reply:
updated